File: src/leetcode-11/go/code.go
package main

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func MaxArea(height []int) int {
	var left, maxVol int
	right := len(height) - 1

	for right > left {
		w := right - left
		h := min(height[left], height[right])
		maxVol = max(w*h, maxVol)
		if height[left] > height[right] {
			right--
		} else {
			left++
		}
	}

	return maxVol
}

File: src/leetcode-11/ts/code.ts
export function maxArea(height: number[]): number {
	let left = 0;
	let right = height.length - 1;
	let waterVolume = 0;
	while (left < right) {
		const w = right - left;
		const h = Math.min(height[left], height[right]);
		waterVolume = Math.max(w * h, waterVolume);
		if (height[left] < height[right]) {
			left += 1;
		} else {
			right -= 1;
		}
	}
	return waterVolume;
}
